pkg/controller/provider/openstack/handler.go

```diff
@@ -74,13 +74,31 @@ func readAuthConfig(c *provider.DNSHandlerConfig) (*clientAuthConfig, error) {
 	if err != nil {
 		return nil, err
 	}
-	username, err := c.GetRequiredProperty("OS_USERNAME", "username")
-	if err != nil {
-		return nil, err
-	}
-	password, err := c.GetRequiredProperty("OS_PASSWORD", "password")
-	if err != nil {
-		return nil, err
+
+	applicationCredentialID := c.GetProperty("OS_APPLICATION_CREDENTIAL_ID", "applicationCredentialID")
+	applicationCredentialName := c.GetProperty("OS_APPLICATION_CREDENTIAL_NAME", "applicationCredentialName")
+	applicationCredentialSecret := c.GetProperty("OS_APPLICATION_CREDENTIAL_SECRET", "applicationCredentialSecret")
+	username := c.GetProperty("OS_USERNAME", "username")
+	password := c.GetProperty("OS_PASSWORD", "password")
+	if applicationCredentialID != "" || applicationCredentialName != "" {
```

Comment:
All this logic is already implemented in [gophercloud package](https://github.com/gophercloud/gophercloud/blob/1bcad6c25756b93ab1f2bd672da6a6ff3574f2be/auth_options.go#L224-L282). Is it really necessary to duplicate the logic?

Reply:
Thanks for the hint!
But I think it makes sense to perform the checks at this location. The properties are read from a secret with alternative key names for historical reasons. For example the username can be provided with key 'OS_USERNAME' or 'username'. It is helpful for locating the misconfiguration if the message for such basic errors contains all possible key names.